backupccl: add TestOnlineRestoreS3 and TestOnlineRestoreBasic tests

msbutler opened the pull request:
This patch adds TestOnlineRestoreS3 to our cloud unit test suite. To run locally, run:
```
./dev test pkg/ccl/backupccl -f TestOnlineRestoreS3 -- --test_env=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID --test_env=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY --test_env=AWS_S3_BUCKET=cockroachdb-backup-cloud-nightly --test_env=COCKROACH_UNSAFE_RESTORE=true
```

This patch also adds TestOnlineRestoreBasic which runs online restore through
nodelocal. To run locally, run:
```
./dev test pkg/ccl/backupccl -f TestOnlineRestoreBasic -- --test_env=COCKROACH_UNSAFE_RESTORE=true
```

Both tests will be skipped in the nightlies until we pass
COCKROACH_UNSAFE_RESTORE to their setup.

Epic: none

Release note: none

msbutler committed:
backupccl: add TestOnlineRestoreS3 and TestOnlineRestoreBasic tests

This patch adds TestOnlineRestoreS3 to our cloud unit test suite. To run locally, run:
```
./dev test pkg/ccl/backupccl -f TestOnlineRestoreS3 -- --test_env=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID --test_env=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY --test_env=AWS_S3_BUCKET=cockroachdb-backup-cloud-nightly --test_env=COCKROACH_UNSAFE_RESTORE=true
```

This patch also adds TestOnlineRestoreBasic which runs online restore through
nodelocal. To run locally, run:
```
./dev test pkg/ccl/backupccl -f TestOnlineRestoreBasic -- --test_env=COCKROACH_UNSAFE_RESTORE=true
```

Both tests will be skipped in the nightlies until we pass
COCKROACH_UNSAFE_RESTORE to their setup.

Epic: none

Release note: none

msbutler committed:
backupccl: fix online restore deadlock

Previously, if sendAddRemoteSSTs would return an error, the online restore
would hang because the coordinator did not close the tracingAggregator channel.
This patch fixes this.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114279)
<!-- Reviewable:end -->

msbutler commented:
Another reason to currently skip these tests. Online Restore fails loudly in pebble:
```
=== RUN   TestOnlineRestoreS3
    test_log_scope.go:170: test logs captured to: /home/michaelbutler/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/95e138e66d69292427dfb9528cf06d04/logTestOnlineRestoreS32969412245
    test_log_scope.go:81: use -show-logs to present logs inline
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1  while ingesting data/916152420942577665.sst: pebble: cannot ingest external file with no point or range keys
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !goroutine 5099 [running]:
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/allstacks.GetWithBuf({0x0?, 0x30?, 0x367300?})
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/pkg/util/allstacks/allstacks.go:32 +0x74
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/allstacks.Get(...)
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/pkg/util/allstacks/allstacks.go:19
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(_, {{{0xc004f16690, 0x24}, {0x643c45d, 0x1}, {0x643c45d, 0x1}, {0x6368603, 0x6}, {0x643c45d, ...}}, ...})
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/pkg/util/log/clog.go:282 +0xc6
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x7af68a0, 0xc006866360}, 0x2, 0x4, 0x0, 0x0?, {0x63c9d65, 0x16}, {0xc0071a7ea0, 0x2, ...})
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/pkg/util/log/channels.go:109 +0x585
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(...)
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 ! github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:848
F231110 23:12:12.753005 5099 kv/kvserver/replica_proposal.go:671  [T1,Vsystem,n1,s1,r67/1:/{Table/65-Max},raft] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.addSSTablePreApply({0x7af68a0, 0xc006866360}, {0xc00455a000, {0x7bc31d0, 0xc0057dfb00}, {0x7b3c888, 0xc006b2b940}, 0xc0075ac690}, 0x0?, 0xd, ...)
```

msbutler committed:
kvserver: assume all external files have point keys in addSSTablePreApply

This hack allows online restore to succeed. In the long run, I guess we'll need
to know if the external sst has point or range keys in the provided span.

Epic: none

Release note: none

msbutler committed:
backupccl: remove online restore env var gate

This patch removes the COCKROACH_UNSAFE_RESTORE env var which gates online
restore. The gate no longer seems necesssary as all development will
occur on master for the next few months. In addition, the env var makes it
slightly annoying to test online restore.

Epic: None

Release note: None

msbutler commented on pkg/kv/kvserver/replica_proposal.go:
```diff
@@ -655,6 +655,10 @@ func addSSTablePreApply(
 			SmallestUserKey: start.Encode(),
 			LargestUserKey:  end.Encode(),
+
+			// TODO(msbutler): I guess we need to figure out if the backing external
```
@itsbilal Given the preso this morning, I guess I should reword this comment to: "we set this true even if the virtual SST's span covers no point keys in the remote sst, to avoid a pebble level panic during ingestion. At download time, if the backing file has no actual keys in the span, the virtual sst will be compacted away and no keys will download.

msbutler commented:
TFTR!

bors r=dt

itsbilal commented on pkg/kv/kvserver/replica_proposal.go:
```diff
@@ -655,6 +655,10 @@ func addSSTablePreApply(
+
+			// TODO(msbutler): I guess we need to figure out if the backing external
+			// file has point or range keys in the target span.
+			HasPointKey: true,
```
It's not possible to have range keys in backup SSTs at the moment right? So this might be sufficient and good enough for us.

itsbilal commented on pkg/server/config.go:
```diff
@@ -802,6 +802,7 @@ func (cfg *Config) CreateEngines(ctx context.Context) (Engines, error) {
 			}
 			addCfgOpt(storage.MaxSize(sizeInBytes))
 			addCfgOpt(storage.CacheSize(cfg.CacheSize))
+			addCfgOpt(storage.RemoteStorageFactory(cfg.ExternalStorageAccessor))
```
I imagine this is test-only code seeing as it's under `InMem`, while for regular engines we already pass this cfg down below.

itsbilal reviewed (approved):
Storage side LGTM
